src/async_fuse/memfs/s3_metadata.rs

```diff
@@ -1094,6 +1076,136 @@ impl<S: S3BackEnd + Send + Sync + 'static> S3MetaData<S> {
         }
     }
 
+    /// Helper function to pre-check for exchange rename.
+    ///
+    /// This function ensures:
+    /// 1. The old parent, old entry, the new parent and the new entry exist.
+    /// 2. Both the old parent and the new parent are directories.
+    /// 3. The user renaming the file is permitted to do this operation when the sticky bit of one of the old entry or the new entry is set.
+    ///
+    /// When all checks above passed,
+    /// this function returns a tuple containing:
+    /// - A `S3Node` of old parent
+    /// - The ino of old entry
+    /// - A `S3Node` of new parent, or `None` if the new parent is same as old parent
+    /// - The ino of new entry
+    ///
+    /// Otherwise, it returns an `Err`.
+    #[cfg(feature = "abi-7-23")]
+    async fn exchange_pre_check<T: MetaTxn + ?Sized>(
+        &self,
+        txn: &mut T,
+        context: &ReqContext,
+        old_parent: INum,
+        old_name: &str,
+        new_parent: INum,
+        new_name: &str,
+    ) -> DatenLordResult<(S3Node<S>, INum, Option<S3Node<S>>, INum)> {
+        let check_sticky_bit = |parent_node: &S3Node<S>, entry: &DirEntry| {
+            let parent_attr = parent_node.get_attr();
+            if context.user_id != 0
+                && (parent_attr.perm & 0o1000 != 0)
+                && context.user_id != parent_attr.uid
+                && context.user_id != entry.file_attr_arc_ref().read().uid
+            {
+                build_error_result_from_errno(Errno::EACCES, "Sticky bit set".to_owned())
+            } else {
+                Ok(())
+            }
+        };
+
+        let check_node_is_dir = |node: &S3Node<S>| {
+            if node.get_type() == SFlag::S_IFDIR {
+                Ok(())
+            } else {
+                build_error_result_from_errno(
+                    Errno::ENOTDIR,
+                    format!("{} is not a dir.", node.get_ino()),
+                )
+            }
+        };
+
+        let build_enoent = |name: &str, parent: INum, parent_node: &S3Node<S>| {
+            build_error_result_from_errno(
+                Errno::ENOENT,
+                format!(
+                    "exchange_pre_check() failed to find child entry of name={:?} \
+                        under parent directory ino={} and name={:?}",
+                    name,
+                    parent,
+                    parent_node.get_name(),
+                ),
+            )
+        };
+
+        let old_parent_node = txn
+            .get(&KeyType::INum2Node(old_parent))
+            .await?
+            .ok_or_else(|| {
+                anyhow::Error::new(Errno::ENOENT)
```

Comment:
Should we use anyhow error here? It's a typical error telling some resource doesn't exist. I think it's a "large" error deserves a datenlord error type.

Reply:
It does deserve a DatenLord error type, but we cannot just return a value of other error types here now.

The codes here are imitating the widely used function `build_error_result_from_errno`, which returns an anyhow error:

```rust
pub fn build_error_result_from_errno<T>(error_code: Errno, err_msg: String) -> DatenLordResult<T> {
    Err(DatenLordError::from(
        anyhow::Error::new(error_code).context(err_msg),
    ))
}
```

The callers (in fact, they are methods of `MemFs` as `FileSystem` implementation) of those methods in `Metadata` will get a returned `DatenLordError`, and call `reply.error()` to send it to the kernel, if it is an `Err`.  The `error()` is defined as below:

```rust
pub async fn error(self, err: DatenLordError) -> nix::Result<usize> {
    self.reply.send_error(err).await
}
```
The `self.reply` is an instance of `ReplyRaw`, and `ReplyRaw::send_error()` is defined as:

```rust
async fn send_error(self, err: DatenLordError) -> nix::Result<usize> {
    match err {
        DatenLordError::InternalErr { source, context } => {
            let error_code = if let Some(nix_err) =
                source.root_cause().downcast_ref::<nix::Error>()
            {
                if *nix_err == nix::errno::Errno::UnknownErrno {
                    panic!(
                        "should not send nix::errno::Errno::UnknownErrno to FUSE kernel, \
                                the error is: {} ,context is : {:?}",
                        crate::common::util::format_anyhow_error(&source),
                        context,
                    );
                } else {
                    nix_err
                }
            } else {
                panic!(
                        "should not send non-nix error to FUSE kernel, the error is: {},context is : {:?}",
                        crate::common::util::format_anyhow_error(&source),context,
                    );
            };
            self.send_error_code(*error_code).await
        }
        err => {
            panic!("should not send non-internal error to FUSE kernel ,the error is : {err}",);
        }
    }
}
```

That's the problem! The `err` parameter must be a `DatenLordError::InternalErr`, with an `anyhow::Error` inside, and must be able to be downcast to an `nix::Errno`, otherwise the application will panic.

That's why the function `build_error_result_from_errno` and I return an anyhow error but not a seperated error. To solve such problems, we may need to refactor the mechanism of error handling of DatenLord.